src/storage/azure/storage.c

```diff
@@ -0,0 +1,729 @@
+
+        // Generate canonical headers
+        String *headerCanonical = strNew("");
+        StringList *headerKeyList = httpHeaderList(httpHeader);
+
+        for (unsigned int headerKeyIdx = 0; headerKeyIdx < strLstSize(headerKeyList); headerKeyIdx++)
+        {
+            const String *headerKey = strLstGet(headerKeyList, headerKeyIdx);
+
+            if (!strBeginsWithZ(headerKey, "x-ms-"))
+                continue;
+
+            strCatFmt(headerCanonical, "%s:%s\n", strPtr(headerKey), strPtr(httpHeaderGet(httpHeader, headerKey)));
+        }
+
+        // Generate canonical query
+        String *queryCanonical = strNew("");
+
+        if (query != NULL)
+        {
+            StringList *queryKeyList = httpQueryList(query);
+            ASSERT(strLstSize(queryKeyList) > 0);
+
+            for (unsigned int queryKeyIdx = 0; queryKeyIdx < strLstSize(queryKeyList); queryKeyIdx++)
+            {
+                const String *queryKey = strLstGet(queryKeyList, queryKeyIdx);
+
+                strCatFmt(queryCanonical, "\n%s:%s", strPtr(queryKey), strPtr(httpQueryGet(query, queryKey)));
+            }
+        }
+
+        // Generate string to sign
+        const String *contentLength = httpHeaderGet(httpHeader, HTTP_HEADER_CONTENT_LENGTH_STR);
+        const String *contentMd5 = httpHeaderGet(httpHeader, HTTP_HEADER_CONTENT_MD5_STR);
+
+        const String *stringToSign = strNewFmt(
+            "%s\n"                                                  // verb
+            "\n"                                                    // content-encoding
+            "\n"                                                    // content-language
+            "%s\n"                                                  // content-length
+            "%s\n"                                                  // content-md5
+            "\n"                                                    // content-type
+            "%s\n"                                                  // date
+            "\n"                                                    // If-Modified-Since
+            "\n"                                                    // If-Match
+            "\n"                                                    // If-None-Match
+            "\n"                                                    // If-Unmodified-Since
+            "\n"                                                    // range
+            "%s"                                                    // Canonicalized headers
+            "/%s%s"                                                 // Canonicalized account/uri
+            "%s",                                                   // Canonicalized query
+            strPtr(verb), strEq(contentLength, ZERO_STR) ? "" : strPtr(contentLength), contentMd5 == NULL ? "" : strPtr(contentMd5),
+            strPtr(dateTime), strPtr(headerCanonical), strPtr(this->account), strPtr(uri), strPtr(queryCanonical));
```

Comment:
We are implementing version `2019-02-02` which uses the latter format.  The version doesn't come from the server -- we tell the server which version we will be using.

There doesn't seem to be a need to implement anything earlier than that since the Azure service and Azurite (the only emulator) both support `2019-02-02`.